### DISCUSSION TITLE
Script injection through data

### DISCUSSION THREAD

**Discussion by hiranpodi** [d3/d3#3516] [Q&A]
What are the mechanises in place in D3 to protect script injection through data?

**Comment by harshagarwalnyu** [comment]
D3 itself has **no built-in sanitization** - it is a data transformation + DOM-binding library, not a security framework. The injection risk depends entirely on which D3 API you use:

| API | Under the hood | Safe? |
|-----|---------------|-------|
| `.text(value)` | `element.textContent = value` | Yes - always treated as text |
| `.attr('class', value)` | `element.setAttribute(...)` | Yes for most attributes |
| `.html(value)` | DOM tree mutation via parsed HTML | **No** - executes embedded scripts in some contexts |
| `.attr('href', value)` | `element.setAttribute('href', ...)` | Risky if value starts with `javascript:` |
| `.on('click', fn)` | `addEventListener` | Yes - takes a function, not a string |

**The main risk: `.html()`**

D3's `.html()` method parses its argument as HTML markup and injects the resulting DOM tree, which can include script elements and event handler attributes. Avoid it when the value comes from user or external data:

```js
// DANGEROUS if `d` comes from untrusted data:
svg.selectAll('text').data(data).join('text').html(d => d.label);

// SAFE: use .text() for plain string content:
svg.selectAll('text').data(data).join('text').text(d => d.label);
```

**SVG `<foreignObject>`** can embed arbitrary HTML - apply the same care.

**Practical rules:**
1. Prefer `.text()` over `.html()` everywhere possible
2. If you must call `.html()` with external data, sanitize first with [DOMPurify](https://github.com/cure53/DOMPurify): `selection.html(DOMPurify.sanitize(d.label))`
3. Validate `href`/`src` attribute values against an allowlist of protocols (`https:`, relative paths)
4. CSP headers (`script-src 'self'`) are your last line of defense and block inline script injection even if sanitization is missed

D3 visualizations driven by fully controlled/static data have no injection risk. The risk surface is purely at the point where untrusted data enters a DOM-mutating D3 call.
